rust/lance-index/src/vector/hnsw/builder.rs

```diff
@@ -733,14 +737,134 @@ impl IvfSubIndex for HNSW {
             storage.distance_type(),
         );
 
+        let chunk_size = hnsw.inner.params.ef_construction;
+
         let len = storage.len();
         hnsw.inner.level_count[0].fetch_add(1, Ordering::Relaxed);
-        (1..len).into_par_iter().for_each_init(
-            || VisitedGenerator::new(len),
-            |visited_generator, node| {
-                hnsw.inner.insert(node as u32, visited_generator, storage);
-            },
-        );
+
+        thread_local! {
+            static VISITED_GENERATOR: RefCell<Option<VisitedGenerator>> = RefCell::new(None);
+        }
+
+        (1..len)
+            .collect::<Vec<_>>()
+            .chunks(chunk_size) // Split the range into chunks of the specified size
+            .for_each(|chunk| {
+                // Phase I: Obtain a clique of candidate edges within each chunk at each level
+                let local_levels: Vec<u16> = chunk
+                    .into_par_iter()
+                    .map(|&node| {
+                        hnsw.inner.nodes[node as usize]
+                            .read()
+                            .unwrap()
+                            .level_neighbors
+                            .len() as u16
+                            - 1
+                    })
+                    .collect();
+                chunk
+                    .into_par_iter()
```

Comment:
This assumes the chunk size is small, not sure whether it can benefit from the parallel. Did you try `par_chunks`? But I guess. that would require more memory

Reply:
See other comment for why this complexity works out. Chunks are intended to be big enough to benefit from parallelism, while being small enough to not change the complexity of the algorithm.